### DISCUSSION TITLE
Changelog support

### DISCUSSION THREAD

**Discussion by delucis** [withastro/starlight#1073] [Feature Requests]
### What version of `starlight` are you using?

0.12

### What is your idea?

Some utility for handling content like changelogs. Might be best suited to a plug-in and maybe a userland solution would work well.

Pagination utilities to build collections like the Starlight blog plug-in creates could be relevant.

### Why is this feature necessary?

Cross posting some user feedback we got:

> Obviously with docs, most of them have a changelog. I'd want something more opinionated around that (changelog).
> 
> For example. A list view that you could show the most recent 10. An archive to paginate through old entries. Maybe each changelog entry would have its own page and a link to that that we can share.

### Participation

- [ ] I am willing to submit a pull request for this feature.

**Comment by HiDeoo** [comment]
We are talking about general changelog and not changelog for the actual documentation, right?

If that is the case, I have internally an integration for that purpose that I plan to rewrite and open source at some point (after #982 is fixed as it's running on older Starlight version at the moment to avoid this issue).

If that's interesting to people, sharing how it works at the moment:

- You get new pages kinda like the blog integration you mentioned with a paginated list of all releases with some filtering options
- You get individual pages for each release with the changelog
- You get either a new sidebar entry for the release section or a group with links to the last X (configurable) releases
- You can optionally not use the sidebar entries and have a link in the header to the release section instead
- There are currently 2 sources possible for the changelog:
  - A repo and the GitHub API
  - Manual content collection entries for repo that don't publish GitHub releases with a new schema (like blog post with the blog integration basically)

**Reply by delucis** [reply]
Nice that sounds fantastic and likely would meet that user’s needs exactly. Only other potential data source I can imagine would be the ability to ingest a `CHANGELOG.md` file and parse out headings for versions into entries somehow.

**Reply by vasfvitor** [reply]
Seeing #982 closed, how is the current state of your integration @HiDeoo ?

**Reply by HiDeoo** [reply]
It definitely took a while, but I finally got around to rewriting it entirely using the Astro’s Content Loader API and [publishing](https://github.com/HiDeoo/starlight-changelogs) it as a Starlight plugin.

The plugin can load changelogs from various sources, e.g. a `CHANGELOG.md` file generated by Changesets, GitHub Releases, and more to come. Pages listing every version with pagination are automatically generated and same for individual version pages. It also supports a feature that I deeply miss from GitHub Releases: the ability to display all changes from a specific version to the latest version, which is useful to see in a single place everything that has changed since the last time you updated.

An example of a generated page can be found [here](<https://starlight-changelogs.netlify.app/demo/starlight-blog/version/0-21-0/>) and the `starlight-changelogs` plugin is now available [here](https://github.com/HiDeoo/starlight-changelogs).

**Comment by JEverhart383** [comment]
Just wanted to chime in here to say that WP Engine would absolutely use some sort of changelog integration here: https://developers.wpengine.com/docs/

I explored @HiDeoo's existing blog integration a few weeks ago and felt like that was pretty close to what we were looking for but put a pin in it for the moment. I imagine many other products are in the same boat and this would be an extraordinarily helpful integration.

**Reply by arimgibson** [reply]
Yup, I was planning on the blog integration as a changelog as well!

**Comment by mayank99** [comment]
Suddenly finding myself thinking about changelogs and release notes a lot recently, and pleasantly surprised to find other people in the same boat too. 😅

I think [`changesets`](https://github.com/changesets/changesets) got a lot of things right, including git being used as the source of truth. It would be neat if Starlight extended the changesets workflow to generate changelog entries in multiple places at once:
- `CHANGELOG.md` inside package(s) source directory
- Github release notes
- Starlight changelog pages! ✨
